parser/printer.go

```diff
@@ -15,351 +15,472 @@
+	if addNewlines {
+		p.getEmptyLine()
 	}
-	p.printToken("]", endPos)
+	p.printString("]")
 }
 
 func (p *printer) printMap(m *Map) {
-	p.requestSpace()
-	p.printToken("{", m.LBracePos)
-	if len(m.Properties) > 0 || m.LBracePos.Line != m.RBracePos.Line {
-		p.requestNewline()
-		p.indent(p.curIndent() + 4)
-		for _, prop := range m.Properties {
-			p.printProperty(prop)
-			p.printToken(",", noPos)
-			p.requestNewline()
-		}
-		p.unindent(m.RBracePos)
+	p.printString("{")
+	p.incrementNextIndent()
+	if m.MapBody.Properties == nil {
+		panic("nil mapBody")
+	}
+	p.printNode(m.MapBody)
+	p.decrementNextIndent()
+	if len(m.MapBody.Properties) > 0 || len(p.syntaxTree.GetComments(m.MapBody).preComments) > 0 || len(p.syntaxTree.GetComments(m.MapBody).postComments) > 0 {
+		p.getEmptyLine()
 	}
-	p.printToken("}", m.RBracePos)
+	p.printString("}")
 }
 
-func (p *printer) printOperator(operator *Operator) {
-	p.printExpression(operator.Args[0])
-	p.requestSpace()
-	p.printToken(string(operator.Operator), operator.OperatorPos)
-	if operator.Args[0].End().Line == operator.Args[1].Pos().Line {
-		p.requestSpace()
-	} else {
-		p.requestNewline()
+// the reason that a MapBody is distinct from a Map is so that comments can be attached before it or after it and still show inside the map, even if the map is empty
+func (p *printer) printMapBody(m *MapBody) {
+	for _, property := range m.Properties {
+
+		// these calls are unwrapped (instead of just doing printNode) to make sure that the trailing comma (after the property) shows up after any inline comments
+		p.getEmptyLine()
+		p.beforePrint(property)
+		p.printString(property.Name)
+		p.printString(":")
+		p.printString(" ")
+
+		// these calls are unwrapped (instead of just doing printNode) to make sure that spacing of inline comments inside the property works correctly
+		p.beforePrint(property.Value)
+		p.getSpace()
+		p.printNodeWithoutComments(property.Value)
+		p.afterPrint(property.Value)
+		p.printString(",")
+
+		p.afterPrint(property)
+		p.getEmptyLine()
 	}
-	p.printExpression(operator.Args[1])
 }
 
-func (p *printer) printProperty(property *Property) {
-	p.printToken(property.Name, property.NamePos)
-	p.printToken(":", property.ColonPos)
-	p.requestSpace()
-	p.printExpression(property.Value)
+func (p *printer) printOperator(operator *Operator) {
+	p.printNode(operator.Args[0])
+	p.requireSpace()
+	p.printString(operator.OperatorToken.Value)
+	p.requireSpace()
+	p.printNode(operator.Args[1])
 }
 
-// Print a single token, including any necessary comments or whitespace between
-// this token and the previously printed token
-func (p *printer) printToken(s string, pos scanner.Position) {
-	newline := p.pendingNewline != 0
-
-	if pos == noPos {
-		pos = p.pos
-	}
-
-	if newline {
-		p.printEndOfLineCommentsBefore(pos)
-		p.requestNewlinesForPos(pos)
+func (p *printer) beforePrint(node ParseNode) {
+	var comments = p.syntaxTree.GetCommentsIfPresent(node)
+	if comments != nil {
```

Comment:
Hmm, when I remove the check, I get a panic trying to access comments.preComments